azure-kusto-data/azure/kusto/data/_models.py

```diff
@@ -48,26 +91,51 @@ def __init__(self, columns, row):
                         # this precision in case they want it. One example why one might want this precision, is when
                         # working with pandas. In that case, use azure.kusto.data.helpers.dataframe_from_result_table
                         # which takes into account the 7th digit.
-                        char = value.split(":")[2].split(".")[1][6]
-                        if char.isdigit():
-                            tick = int(char)
-                            last = value[-1] if value[-1].isalpha() else ""
-                            typed_value = KustoResultRow.convertion_funcs[lower_column_type](value[:-2] + last)
+                        seconds_part = value.split(":")[2]
+                        seconds_fractions_part = seconds_part.split(".")[1]
+                        seventh_char = seconds_fractions_part[6]
+
+                        if seventh_char.isdigit():
+                            tick = int(seventh_char)
+                            lookback = 2 if last else 1
+                            typed_value = KustoResultRow.convertion_funcs[column_type](value[:-lookback] + last)
+
                             if tick:
-                                if lower_column_type == "datetime":
+                                if column_type == "datetime":
                                     self._seventh_digit[column.column_name] = tick
-                                else:
+                                elif column_type == "timespan":
                                     self._seventh_digit[column.column_name] = (
                                         tick if abs(typed_value) == typed_value else -tick
                                     )
+                                else:
```

Comment:
I am not sure that this raise is necessary, since we check the types at line 74.
But it is your call